libsrc/serialization.cpp

```diff
@@ -0,0 +1,214 @@
+#include <sstream>
+#include <string>
+
+#include "ismrmrd/serialization.h"
+
+namespace ISMRMRD {
+
+void serialize(const Acquisition &acq, WritableStream &ws) {
+    AcquisitionHeader ahead = acq.getHead();
+    ws.write(reinterpret_cast<const char *>(&ahead), sizeof(AcquisitionHeader));
+    ws.write(reinterpret_cast<const char *>(acq.getTrajPtr()), ahead.trajectory_dimensions * ahead.number_of_samples * sizeof(float));
+    ws.write(reinterpret_cast<const char *>(acq.getDataPtr()), ahead.number_of_samples * ahead.active_channels * 2 * sizeof(float));
+}
+
+template <typename T>
+void serialize(const Image<T> &img, WritableStream &ws) {
+    ImageHeader ihead = img.getHead();
+    if (ismrmrd_sizeof_data_type(ihead.data_type) != sizeof(T)) {
+        throw std::runtime_error("Image data type does not match template type");
+    }
+    ws.write(reinterpret_cast<const char *>(&ihead), sizeof(ImageHeader));
+    uint64_t attr_length = img.getAttributeStringLength();
+    ws.write(reinterpret_cast<char *>(&attr_length), sizeof(uint64_t));
+    if (attr_length) {
+        ws.write(img.getAttributeString(), ihead.attribute_string_len);
+    }
+    ws.write(reinterpret_cast<const char *>(img.getDataPtr()), img.getDataSize());
+}
+
+void serialize(const Waveform &wfm, WritableStream &ws) {
+    ws.write(reinterpret_cast<const char *>(&wfm.head), sizeof(ISMRMRD_WaveformHeader));
+    ws.write(reinterpret_cast<const char *>(wfm.begin_data()), wfm.head.number_of_samples * wfm.head.channels * sizeof(uint32_t));
+}
+
+void deserialize(Acquisition &acq, ReadableStream &rs) {
```

Comment:
The deserialize functions are not taking into account a premature EOF, and silently return invalid data.

Reply:
I have added an `eof()` on the readable view interface and `bad()` on the writable stream view. They correspond to the relevant functions on iostream. You would have to implement those if you wrap something else.